mangaki/mangaki/utils/recommendations.py

```diff
@@ -5,119 +5,143 @@
+    random.shuffle(anonymous_u)
+    random.shuffle(anonymous_w)
+    encode_user = dict(zip(users, anonymous_u))
+    encode_work = dict(zip(works, anonymous_w))
+    decode_user = dict(zip(anonymous_u, users))
+    decode_work = dict(zip(anonymous_w, works))
+
+    interesting_works = set()
+    for work_id, _ in nb_ratings.most_common():  # work_id are sorted by decreasing number of ratings
+        if nb_ratings[work_id] < RATED_BY_AT_LEAST:
+            break
+        interesting_works.add(work_id)
+
+    for user_id, work_id, rating in triplets:
+        X.append((encode_user[user_id], encode_work[work_id]))
+        y.append(rating_values[rating])
+    return np.array(X), np.array(y), len(users), len(works), encode_user, decode_user, encode_work, decode_work, interesting_works
+
+
+def get_reco_algo(user, algo_name='knn', category='all'):
+    chrono = Chrono(CHRONO_ENABLED)
 
+    already_rated_works = Rating.objects.filter(user=user).values_list('work_id', flat=True)
 
-NB_NEIGHBORS = 15
-MIN_RATINGS = 3
+    chrono.save('[%dQ] get rated works' % len(connection.queries))
 
-CHRONO_ENABLED = False
+    if algo_name == 'knn':
+        queryset = Rating.objects.filter(work__in=already_rated_works)
+        ratings_pack = make_anonymous_data(queryset)
 
+        chrono.save('[%dQ] make first anonymous data' % len(connection.queries))
 
-def get_recommendations(user, category, editor='unspecified', dpp=False):
-    # What if user is not authenticated? We will see soon.
-    chrono = Chrono(CHRONO_ENABLED)
+        X, y, nb_users, nb_works, encode_user, decode_user, encode_work, decode_work, interesting_works = ratings_pack
+        algo = MangakiKNN(NB_NEIGHBORS)
+        algo.set_parameters(nb_users, nb_works)
+        algo.fit(X, y)
+
+        chrono.save('[%dQ] prepare first fit' % len(connection.queries))
 
-    chrono.save('[%dQ] begin' % len(connection.queries))
+        encoded_neighbors = algo.get_neighbors([encode_user[user.id]])[0]
+        neighbors = [decode_user[encoded_user_id] for encoded_user_id in encoded_neighbors]
 
-    rated_works = {}
-    if dpp:
-        qs = ColdStartRating.objects.filter(user=user).values_list('work_id', 'choice')
-        values = rating_values_dpp
+        chrono.save('[%dQ] get neighbors (checksum: %d)' % (len(connection.queries), sum(neighbors)))
 
+        queryset = Rating.objects.filter(user__id__in=neighbors + [user.id]).exclude(choice__in=['willsee', 'wontsee'])
     else:
-        qs = Rating.objects.filter(user=user).values_list('work_id', 'choice')
-        values = rating_values
-
-    for work_id, choice in qs:
-        rated_works[work_id] = choice
-
-    willsee = set()
-    if user.profile.reco_willsee_ok:
-        banned_works = set()
-        for work_id in rated_works:
-            if rated_works[work_id] != 'willsee':
-                banned_works.add(work_id)
-            else:
-                willsee.add(work_id)
+        queryset = Rating.objects.all()
```

Comment:
Why do we get *ALL* the ratings if we are not using a KNN?
(Moreover: If we could do a `if not knn` rather than a big `if`, that'd be more readable.)